rfcs/rdm-0068-data-migration.md

```diff
@@ -0,0 +1,369 @@
+
+## Future work / Unresolved Questions
+
+### State handling
+
+The current implementation has no state handling. Data is piped through the
+stream and will either succeed or fail, but there is no observability or
+checkpointing. This is also due to the fact that the staging database is
+_memory_.
+
+Below is defined a potential set of states that would solve this problem:
+
+- Default
+    - status: `TRANSFORMATION_INIT`.
+- Transform:
+    - initial: `TRANSFORMATION_INIT`.
+    - if fails, `TRANSFORMATION_FAILED`.
+    - if succeed, `TRANSFORMATION_SUCCEEDED`.
+- Load:
+    - initial: `VALIDATION_INIT`.
+    - if validation fails, `VALIDATION_FAILED`. Once the validation error is
+    identified (manual intervention), for example wrong transformation, status
+    `TRANSFORMATION_INIT`.
+    - if validation succeeds, `LOAD_INIT`, the record is valid and can be loaded
+    in the new InvenioRDM database.
+    - if loading fails, `LOAD_FAIL`, after identifying the issue if it needs
+    re-processing then `TRANSFORMATION_INIT`.
+    - if loading succeds, `LOAD_SUCCEED`. Records will be marched as loaded.
+
+| layer     | status                | original record | transformed record | valid |
+|:--------- |:--------------------- |:--------------- | ------------------ | ----- |
+| Transform | `TRANSFORMATION_INIT` | original        | null               | no    |
+| Transform | `TRANSFORMATION_FAIL` | original        | null               | no    |
+| Load      | `VALIDATION_INIT`     | original        | transformed record | no    |
+| Load      | `VALIDATION_FAIL`     | original        | transformed record | no    |
+| Load      | `LOAD_INIT`           | original        | transformed record | yes   |
+| Load      | `LOAD_FAIL`           | original        | transformed record | yes   |
+| Load      | `LOAD_SUCCEED`        | original        | transformed record | yes   |
+
+Note that all states will contain the original and current state of the record,
+this data lineage gives the needed flexibility to re-process when needed and the
+observability to know what failed.
+
+### Monitoring
+
+Monitoring is needed at task level, meaning how many items are being process
+at the moment, how many are left, etc. and also at application level, to have
+details in case of failures (e.g. transformer). The second one could be achieved
+by connecting a Sentry instance and configuring the Python logger.
+
+### Persistance
+
+Using _memory_ as a staging database means that any intermediate state is lost
+in case of errors.
+
+### Checkpointing
+
+Tasks cannot be resumed from where they failed since no state is saved.
+
+### Dependencies between streams
+
+Streams are run sequentially, therefore the order can be enforced when
+configured.
+
+#### Referential validation
+
+The validation step on `Load` would ideally be run on the destination InvenioRDM
+instance, by pulling data from the staging database. At this step data could be
```

Comment:
```suggestion
instance, by pulling data from the staging database. At this step data could be
```

staging database is described in line 311 as non persistent, how can we pull the data from the memory then?

Reply:
It is described as "potentially" not persistence, which is what we have now. Technically, that happens when we store things in Python list/dict. Then when we _read_ it we are _pulling_ data. Maybe _read_ would be a better phrasing?
```suggestion
instance, by reading data from the staging database. At this step data could be
```